### DISCUSSION TITLE
Fluxcd Helm reconciliation info

### DISCUSSION THREAD

**Discussion by mcammisa78** [fluxcd/flux2#1972] [Q&A] [answered]
Hi, I'm evaluating to introduce fluxcd into our company, in particular we are interested in fluxcd helm CD and the reconciliation capability.
After a couple of test we discover that the reconciliation on hr is working fine only for changes introduced manually by Helm cli for example, but if we use oc/kubectl or the cluster console for changing kubernetes object directly deployed thru fluxcd , the reconciliation does not work.
We were aspecting something similar to Operator reconciliation. But it seems not. Is that correct? is it an aspected behaviour?
or  are we doing something wrong ? how reconciliation works for helm chart ? are fluxcd able to check changes applyed thru openshift console directly on a config map that is part of Helm chart for example? I would really appreciate any help

**Comment by makkes** [comment]
Great question. So if you have a HelmRelease that e.g. installs a Helm chart that creates a ConfigMap and then you modify that ConfigMap then Flux won't do anything about it. Flux (in particular the helm-controller) doesn't watch the resources created by a chart but only the HelmRelease itself. So if you e.g. update the version of a chart in the HelmRelease then Flux will upgrade the release and your modifications to said ConfigMap would be overwritten.

Does that answer your question?

**Reply by stefanprodan** [reply]
if you run `helm upgrade` for a release managed by helm-controller, the controller will detect the drift in the Helm storage and undo the changes. If you bypass Helm and change the underlying resources with kubectl, helm-controller can’t detect that. Only kustomize-controller detects changes to Kubernetes resources as it manages them directly.

**Reply by mcammisa78** [reply]
Exactly, we were thinking to write a simple Helm controller able to reconcile the "rendered" kubernetes resources deployed by Helm (like the flux kustomize controller), probably comparing the deployed kube resources with something generated thru Helm cli template command with the current values available on the cluster, the reconciliation will not be performed thru helm upgrade/install but directly applying the single manifest that is changed. In this way the helm revision will not be increased.

Our deploy pipelines are already implementing something similar in order to notify the deployer that something is changed on cluster side so he can get that change and apply it back to the values stored in a segregated git repo. 

Thru something like:
`helm template -n mynamespace --no-hooks --skip-tests -f helm/deploy-values.yaml myapp helm/mychart | kubectl diff --namespace derps -f - | bat -l diff -`

Flux is a more completed solution that include several other features that we are evaulating. As per the below comment it seems that it will be supported by helm controller.

**Comment by hiddeco** [comment, marked as answer]
This is expected behavior at the moment, see: https://github.com/fluxcd/helm-controller/issues/186

The plan is however to support this in the future (https://github.com/fluxcd/helm-controller/issues/186#issuecomment-932107655), but it is blocked by a bigger unit of work that should solve some other observation issues first.

**Reply by mcammisa78** [reply]
I'll will follow the progess of issue 186. Thanks
